Block bindings: Add support for image caption

SantosGuillamot opened the pull request:
Trac ticket: https://core.trac.wordpress.org/ticket/61466

This pull request should be tested together with [this other one from Gutenberg](https://github.com/WordPress/gutenberg/pull/61255), because it needs changes in the image render file.

## What?

Add support for the image caption attribute in block bindings.


## Why?

There is an issue with pattern overrides caused by this: https://github.com/WordPress/gutenberg/issues/62287.

Additionally, it is a good enhancement.

## How?

I created an anonymous class to extend the tag processor and include `set_inner_text` until there is a similar method provided by default.

Additionally, in Gutenberg pull request, I'm modifying the render file of the image to:
* Remove the `figcaption` element when it exists, and the binding value is empty.
* Add the `figcaption` element when it doesn't exist, and the binding value is not empty.

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

SantosGuillamot commented on src/wp-includes/class-wp-block.php:
```diff
@@ -333,7 +333,44 @@ private function replace_html( string $block_content, string $attribute_name, $s
 			case 'html':
 			case 'rich-text':
-				$block_reader = new WP_HTML_Tag_Processor( $block_content );
+				// Create private anonymous class until the HTML API provides `set_inner_html` method.
```
@gziolo @dmsnell I extended the `WP_HTML_Tag_Processor` class this way to include a private `set_inner_text` method until this is possible with the core API. I made this to avoid using a regex. I couldn't find a similar approach in the code. Is this acceptable or should I look for an alternative approach?

dmsnell reviewed (commented):
@SantosGuillamot this is a fine use of sub-classing, though I do want us to be extra careful here, as by adding subclassing code into Core we are setting an example to others, and this is not the example we want to send more broadly, especially since it's coming later.

To that end there are a few things to start with on this. I hope it's okay to extra resilient here.

 - There's no need to create the anonymous class outside of the `if ( 'core/image' … ) {}` block. It's probably fine, but since all of the processing happens within, its scope might be more clear if left inside of it instead of outside of it.

 - let's follow all of Core's documentation requirements, such as a docblock for the added method.

 - one thing I think is essential for "exceptional cases" like this in Core is full checking to ensure we're not trying to replace the wrong tokens. for example, here we might check that the processor is paused on an opening tag. it would be ideal if we also put a terse but scary warning right at the top indicating that this is a stop-gap measure in Core not to be emulated.

```php
public function set_inner_text( $raw_text_content ) {
	if (
		WP_HTML_Tag_Processor::STATE_MATCHED_TAG !== $this->parser_state ||
		$this->is_tag_closer()
	) {
		return false;
	}
}
```

 - It might be ideal to wait to grab the bookmark until the end. The greater the distance between binding the bookmark offsets and its use, the greater the danger. That is, it's fine to set the bookmark at the top, but we shouldn't read it until we add the `lexical_updates[]`.

 - This one isn't important, but since this processor dies immediately we don't need to release the bookmarks.

 - The comment on "Visit the closing tag" is nice, but I think it could really help to indicate that this is a best-effort guess and that code should rely on the HTML Processor for this kind of operation.

To that end, we expect to get image blocks here don't we? Is there anything in the image block that the HTML Processor doesn't support? Maybe we should start instead with it and see if we can skip the Tag Processor.

SantosGuillamot commented:
Thanks a lot for all the feedback, @dmsnell ! That's really helpful 🙂 I totally agree we should be really careful, so please share any more suggestions you have. Going to your points:

> To that end, we expect to get image blocks here don't we? Is there anything in the image block that the HTML Processor doesn't support? Maybe we should start instead with it and see if we can skip the Tag Processor.

Yes, I think it should be safe to use the HTML Processor. I've just changed the code to extend that class instead of the Tag Processor: [link](https://github.com/WordPress/wordpress-develop/pull/6838/files#diff-c569b1ecb11a007a563788a481710b51e0cb21cd1d1885490bf9c91777842f23R338).

I had to pass the `WP_HTML_Processor::CONSTRUCTOR_UNLOCK_CODE` to skip the warning message. I hope that's okay.

> There's no need to create the anonymous class outside of the if ( 'core/image' … ) {} block.

I've moved it into the conditional now. At first, I kept it outside because I was considering using it for other blocks like the paragraph, heading, or button. But we can move it out if we decide to take that approach.

> let's follow all of Core's documentation requirements, such as a docblock for the added method.

I added a comment to the function [here](https://github.com/WordPress/wordpress-develop/pull/6838/files#diff-c569b1ecb11a007a563788a481710b51e0cb21cd1d1885490bf9c91777842f23R339-R350). I added a warning notice trying to push back users from replicating that code. Let me know if that's what you had in mind.

> we might check that the processor is paused on an opening tag.

I added this check as well: [link](https://github.com/WordPress/wordpress-develop/pull/6838/files#diff-c569b1ecb11a007a563788a481710b51e0cb21cd1d1885490bf9c91777842f23R353-R358).

> It might be ideal to wait to grab the bookmark until the end.

I moved this logic just before using lexical updates as suggested: [link](https://github.com/WordPress/wordpress-develop/pull/6838/files#diff-c569b1ecb11a007a563788a481710b51e0cb21cd1d1885490bf9c91777842f23R385-R392).

> since this processor dies immediately we don't need to release the bookmarks

I've removed the bookmarks now.

> I think it could really help to indicate that this is a best-effort guess and that code should rely on the HTML Processor for this kind of operation.

I added another comment trying to clarify this: [link](https://github.com/WordPress/wordpress-develop/pull/6838/files#diff-c569b1ecb11a007a563788a481710b51e0cb21cd1d1885490bf9c91777842f23R368-R371).

---
Please let me know further improvements!

luisherranz commented on src/wp-includes/class-wp-block.php:
```diff
@@ -333,6 +333,73 @@ private function replace_html( string $block_content, string $attribute_name, $s
+						 * USE CASE UNTIL THE HTML PROCESSOR PROVIDES ITS OWN METHOD.
+						 *
+						 * @since 6.6.0
+						 *
+						 * @param string $new_content New text to insert in the HTML element.
+						 * @return bool Whether the inner text was properly replaced.
+						 */
+						public function set_inner_text( $new_content ) {
+							/*
+							 * THIS IS A STOP-GAP MEASURE NOT TO BE EMULATED.
+							 *
+							 * Check that the processor is paused on an opener tag.
+							 *
+							 */
+							if (
+								WP_HTML_Processor::STATE_MATCHED_TAG !== $this->parser_state ||
+								$this->is_tag_closer()
+							) {
+								return false;
+							}
```
I would rename this to `set_content_between_figcaption_balanced_tags` to discourage users from copying this method, and ensure it only processes it if the tag is a `figcaption`.

```suggestion
						/**
						 * Replace the inner content of a figcaption element with the passed content.
						 *
						 * THIS IS A TEMPORARY SOLUTION IN CORE NOT TO BE EMULATED.
						 * IT IS A TEMPORARY SOLUTION THAT JUST WORKS FOR THIS SPECIFIC
						 * USE CASE UNTIL THE HTML PROCESSOR PROVIDES ITS OWN METHOD.
						 *
						 * @since 6.6.0
						 *
						 * @param string $new_content New content to insert in the figcaption element.
						 * @return bool Whether the inner content was properly replaced.
						 */
						public function set_content_between_figcaption_balanced_tags( $new_content ) {
							/*
							 * THIS IS A STOP-GAP MEASURE NOT TO BE EMULATED.
							 *
							 * Check that the processor is paused on an opener tag.
							 *
							 */
							if (
								WP_HTML_Processor::STATE_MATCHED_TAG !== $this->parser_state ||
                'figcaption' !== $this->get_tag() ||
								$this->is_tag_closer()
							) {
								return false;
							}
```

cbravobernal replied:
I think we may need in the future this function not only for captions, but also for more attributes 😅 
Maybe `set_connected_content_between_balanced_tags` or something like that?

SantosGuillamot replied:
Based on the same logic applied in [this comment](https://github.com/WordPress/wordpress-develop/pull/6838#issuecomment-2175880633), where I moved the logic inside the conditional to check it is an image block and caption attribute, I've pushed the commit with the suggestion. We can always revisit the logic if/when using it for more blocks.

ellatrix commented:
I'm quite worried about introducing caption support this late in the release cycle. Imo it deserves some testing as a feature in the GB plugin first. I've made this alternative PR that disallows overrides for an image with caption: https://github.com/WordPress/gutenberg/pull/62747.

SantosGuillamot commented:
I totally understand the concerns. And thanks a lot for the feedback. I agree that if support for image caption is not added, code to disallowing overrides in images with captions for 6.6 could be a good alternative.

dmsnell commented on src/wp-includes/class-wp-block.php:
```diff
@@ -333,6 +333,74 @@ private function replace_html( string $block_content, string $attribute_name, $s
 			case 'rich-text':
+				if ( 'core/image' === $this->name && 'caption' === $attribute_name ) {
+					// Create private anonymous class until the HTML API provides `set_inner_html` method.
+					$bindings_processor = new class( $block_content, WP_HTML_Processor::CONSTRUCTOR_UNLOCK_CODE ) extends WP_HTML_Processor {
```
this is an interesting side-effect of that choice. I think we should rename that constant to something scarier, like what its value is.

I kind of like that it makes anonymous classes stand out more.

dmsnell replied:
@sirreal I think this is a really good example of a part of the HTML Processor we're going to have to improve.

not only is the unlock code ugly here (ugly, because it shouts how this is work-around territory), but also by calling the constructor directly we bypass setting up the context node and the HTML node. in effect, this is starting as a full parser instead of a fragment parser.

I wonder how we can better allow subclassing without this hassle and risk.

dmsnell commented on src/wp-includes/class-wp-block.php:
```diff
@@ -333,6 +333,74 @@ private function replace_html( string $block_content, string $attribute_name, $s
+						 *
+						 * THIS IS A TEMPORARY SOLUTION IN CORE NOT TO BE EMULATED.
+						 * IT IS A TEMPORARY SOLUTION THAT JUST WORKS FOR THIS SPECIFIC
+						 * USE CASE UNTIL THE HTML PROCESSOR PROVIDES ITS OWN METHOD.
```
"Temporary" is a funny word in codebases, as it tends not to be, and it doesn't provide information for future decisions. I like the part indicating that this should be replaced in the future by the HTML Processor's own method.

Maybe we can emphasize this mor along those lines.

> DO NOT COPY THIS METHOD
> THE HTML PROCESSOR WILL HAVE A PROPER METHOD
> USE IT INSTEAD

dmsnell commented on src/wp-includes/class-wp-block.php:
```diff
@@ -333,6 +333,74 @@ private function replace_html( string $block_content, string $attribute_name, $s
+								)
+							) || ! $this->is_tag_closer() ) {
+								return false;
+							}
```
with the HTML Processor we can put away all mention about "balanced tag" and the like. the system is different - there's no more hacking around the thing that is proper.

a fun fact about the recent push/pop system is that when we reach a point where a node closes, the stack of open elements returns the node that opened it.

```php
public function wait_for_the_real__set_inner_html( $theres_a_reason_not_to_do_this ) {
	/*
	 * It should only report a tag name when matched on a tag, so this
	 * is probably enough and we can skip the parser state check.
	 */
	if ( 'FIGCAPTION' !== $this->get_tag() || $this->is_tag_closer() ) {
		return false;
	}

	// Once this element closes the depth will be one shallower than it is now.
	$depth = $this->get_current_depth();
	while ( $this->next_token() && $this->get_current_depth() >= $depth ) {
		// This is inside the FIGCAPTION element.
	}

	if ( null !== $this->get_last_error() || $this->paused_at_incomplete_token() ) {
		return false;
	}

	$this->set_bookmark( 'here' );
	
	$opening = $this->bookmarks[ $this->current_element->token->bookmark_name ];
	$closing = $this->bookmarks[ '_here' ];
	$start   = $opening->start + $opening->length;

	$this->lexical_updates[] = new WP_HTML_Text_Replacement(
		$start,
		$closing->start - $start,
		wp_kses_post( $theres_a_reason_not_to_do_this )
	};

	return true;
}

$block_reader = $bindings_processor::create_fragment( $block_content );
if ( $block_reader->next_tag( 'figcaption' ) ) {
	$block_reader->wait_for_the_real__set_inner_html( $source_value );
}
return $block_reader->get_updated_html();
```

I've also moved the `wp_kses_post()` call into the processor so it's not exposing the unsafe interface by default. we don't want to rely on people remembering or knowing to do the right thing to avoid the dangerous defaults.

there could be some issues with this; hopefully not, but we've been working to ensure that the depth and the breadcrumbs properly report in the present of virtual nodes. for FIGURE and IMG and FIGCAPTION elements though it's unlikely to find invalid markup. that's way more of a problem for P and TABLE and the like.

just let me know if you find any problems: we can work on it together to make sure it works for you.

SantosGuillamot replied:
Thanks a lot for the feedback, @dmsnell  🙂 

I've been testing the suggested logic and it seems to work properly. I just made a commit with it. The only caveat  is that `current_element` is a private property, so I had to make it public. Is this okay or should I look for an alternative method?

dmsnell replied:
I'm glad this worked out for you.

regarding `current_element` I'd like to find a way that doesn't involve opening that up to `public`. the most obvious way is simply to create the bookmark on the `FIGCAPTION` opening tag. I want to end up using the trick that the closer references the opener, but maybe that will have to remain inside the class.

for now, can you try setting an `opener` and `closer` bookmark and reference those instead of the `current_element`?

SantosGuillamot replied:
> can you try setting an opener and closer bookmark and reference those instead of the current_element?

Okay, I've just reverted the `current_element` change and used an `opener` bookmark. Let me know if that's what you had in mind.

dmsnell replied:
if it works for you it works for me!

SantosGuillamot committed:
Clarify comments

SantosGuillamot committed:
Add comments to clarify the set_inner_text function

SantosGuillamot committed:
Add unit test for processing

SantosGuillamot committed:
Move anonymous class inside image conditional

SantosGuillamot committed:
Restrict method to figcaption

Co-authored-by: Luis Herranz <luisherranz@gmail.com>

SantosGuillamot committed:
Use HTML_Processor instead of HTML_Tag_Processor

SantosGuillamot committed:
Fix figcaption conditional

SantosGuillamot committed:
Don't grab bookmarks until the end

SantosGuillamot committed:
Remove release_bookmark calls

SantosGuillamot committed:
Add checks to ensure it doesn't replace unwanted html

SantosGuillamot committed:
Call function properly

SantosGuillamot committed:
Add support for image caption

SantosGuillamot committed:
Update private method comment

SantosGuillamot committed:
Use alternative method

dmsnell commented on src/wp-includes/class-wp-block.php:
```diff
@@ -333,6 +333,60 @@ private function replace_html( string $block_content, string $attribute_name, $s
+							$this->lexical_updates[] = new WP_HTML_Text_Replacement(
+								$start,
+								$closing->start - $start,
+								wp_kses_post( $new_content )
```
if inner text is all we care about we can use `esc_html()` instead of `wp_kses_post()` and that would be a bit more reliable, but if we tried to do something like add a link inside the caption then it will be escaped as `<a href=…>` instead of rendering as a link.

SantosGuillamot replied:
I just checked and you can actually add a link in a caption, so I assume we should keep using `wp_kses_post()`. I have renamed the function to say html instead of text.

SantosGuillamot committed:
Set opening bookmark

SantosGuillamot committed:
Change comments

SantosGuillamot committed:
Change function name

dmsnell commented on src/wp-includes/class-wp-block.php:
```diff
@@ -333,6 +333,64 @@ private function replace_html( string $block_content, string $attribute_name, $s
+						}
+					};
+
+					$block_reader = $bindings_processor::create_fragment( $block_content );
```
whoah. this is unrelated to your changes @SantosGuillamot, but this line really confuses me, purely based on how PHP's anonymous classes work. @sirreal maybe what I wrote above is fine, and when we call `new class( ...$args )` they don't matter?

yes, I just tested and this is odd. calling `$bindings_processor->next_token()` will crash because it never finds the context node. this means that it looks like an HTML Processor but isn't. it's better as a _HTML Processor Builder_.

@SantosGuillamot I wonder if we could take advantage of this in line 338 and instead of passing `$block_content` pass something like `"do not use this, it will not work. it's only here to create a subclass and call the static creator method"`

SantosGuillamot replied:
I am not 100% this is what you meant, but I made [this change](https://github.com/WordPress/wordpress-develop/pull/6838/commits/825f78effbbe234ef70a3502eeaad78ee833436c) to address the feedback. Let me know if that's not what you were suggesting.

dmsnell commented on src/wp-includes/class-wp-block.php:
```diff
@@ -333,6 +333,64 @@ private function replace_html( string $block_content, string $attribute_name, $s
 			case 'rich-text':
+				if ( 'core/image' === $this->name && 'caption' === $attribute_name ) {
+					// Create private anonymous class until the HTML API provides `set_inner_html` method.
+					$bindings_processor = new class( $block_content, WP_HTML_Processor::CONSTRUCTOR_UNLOCK_CODE ) extends WP_HTML_Processor {
```
given the way this works in PHP it could be useful to rename this something like `$bindings_processor_builder`

SantosGuillamot replied:
Changed it as part of [this commit](https://github.com/WordPress/wordpress-develop/pull/6838/commits/825f78effbbe234ef70a3502eeaad78ee833436c).

ellatrix commented on src/wp-includes/class-wp-block.php:
```diff
@@ -333,6 +333,64 @@ private function replace_html( string $block_content, string $attribute_name, $s
 		switch ( $block_type->attributes[ $attribute_name ]['source'] ) {
 			case 'html':
 			case 'rich-text':
+				if ( 'core/image' === $this->name && 'caption' === $attribute_name ) {
```
What is the long term plan here? Are we going to keep making exceptions for some blocks and some attributes? Imo, a more long term solution should be explored, for example templates in the block API, so the server knows how to serialise the block.

SantosGuillamot committed:
Change processor name and args

dmsnell reviewed (commented):
Thanks @SantosGuillamot for all your adjustments. I think this is worth giving it a try. I'd prefer the name of the method be scarier, like `seriously_wait_for_core_to_replace_inner_html()` but that's just a personal preference. This will be I think the first subclass going into Core doing dangerous operations; it makes me nervous is all.

SantosGuillamot commented:
> I'd prefer the name of the method be scarier, like seriously_wait_for_core_to_replace_inner_html() but that's just a personal preference.

While I agree we could change the name to prevent users even more from copying it, `seriously_wait_for_core_to_replace_inner_html` seems too much to me. Here are some other options that, combined with[ the warning message](https://github.com/WordPress/wordpress-develop/pull/6838/files#diff-c569b1ecb11a007a563788a481710b51e0cb21cd1d1885490bf9c91777842f23R345-R347), I believe should be scary enough:

* `do_not_copy_set_figcaption_inner_html`.
* `private_set_figcaption_inner_html`.
* `internal_do_not_use_figcaption_inner_html`.

Apart from that, we could always wait to support the image caption until the HTML Processor provides its own method.

Any thoughts?

cbravobernal commented:
> * `private_set_figcaption_inner_html`

I would go with `private_set_figcaption_inner_html`

dmsnell commented:
What's the status on this? it seems reasonable at the moment, unless there are concerns still about it coming late in the cycle.

SantosGuillamot commented:
> What's the status on this? it seems reasonable at the moment, unless there are concerns still about it coming late in the cycle.

It wasn't considered a priority, and I personally didn't find time to push it. It would probably need a small refactor and testing in Gutenberg as well: [link](https://github.com/WordPress/gutenberg/pull/61255).

Taking into account that the last Gutenberg version included in 6.7 beta will be on Sep 18th, I personally don't have the bandwidth to work on it, but I'm happy to help if needed.

sirreal commented:
@SantosGuillamot and I did some work on an implementation leveraging the HTML processor and using a `set_inner_html` implementation in https://github.com/WordPress/wordpress-develop/pull/7742.
